### PR TITLE
add type traits to be used internally with FEEValuation

### PR DESCRIPTION
**Pull request by davydden**
fixes https://github.com/dealii/dealii/issues/7684

TODO:

- [x] use this with `std::enable_if` to choose between different implementations.

@kronbichler FYI

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3409,8 +3400,98 @@ FEEvaluationBase<dim, n_components_, Number, is_face>::read_cell_data(
...
+  };
+
+
+  // type trait for vector T and Number to see if
+  // we can do vectorized load/save.
+  // for VectorReader and VectorDistributorLocalToGlobal we assume that
+  // if both begin() and local_element()
+  // exist, then begin() + offset == local_element(offset)
+  template <typename T, typename Number>
+  struct vectorizable
```
Comment (jppelteret) [inline include/deal.II/matrix_free/fe_evaluation.h]: Maybe this should be called `is_vectorizable`?

**Comment by masterleinad** [comment]
There are a number of failing tests:
```
	6356 - matrix_free/faces_value_optimization.debug (Failed)
	6357 - matrix_free/faces_value_optimization.release (Failed)
	6358 - matrix_free/faces_value_optimization_02.debug (Failed)
	6359 - matrix_free/faces_value_optimization_02.release (Failed)
	6538 - matrix_free/matrix_vector_faces_09.debug (Failed)
	6539 - matrix_free/matrix_vector_faces_09.release (Failed)
```

**Comment by davydden** [comment]
> There are a number of failing tests:

Interesting, will have a look...

**Comment by davydden** [comment]
Thanks for pushing fixes @masterleinad . A bit surprised why we need a new const accessor.

**Comment by davydden** [comment]
> A bit surprised why we need a new const accessor.

That has to do with `Vector<number>`, looks like a different code path is being used now...

**Comment by davydden** [comment]
@masterleinad I pushed an alternative fix and overwrote for now your commits (I kept a local copy if need them). One might need those changes eventually, but as of this version everything should work without them, I wanted to keep the changes to minimum to ease the review.

Essentially, the question is how to decide whether a vector supports vectorizable read/write? As of now I do it based on existence of `begin()` and `local_element()` for parallel vectors (and matching number types, of course).
 
```
has_begin<T>::value &&
      (has_local_element<T>::value || is_serial_vector<T>::value) &&
      std::is_same<typename T::value_type, Number>::value
```

My interpretation here is that existence of `local_element` implies that the vectors is partitioned according to MPI::Utilities::Partitioner, i.e. `[0, n_loc) + [n_loc, n_loc + n_ghost)`. 

I am ok with this solution (not WIP for now). 
Alternative could be a dedicated type trait similar to `is_serial_vector`. But I don't see to much of an advantage in doing so.

**Comment by davydden** [comment]
@masterleinad i pushed your fixes here as well, but we already see that previous commit was "green" for tests.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -3409,8 +3400,114 @@ FEEvaluationBase<dim, n_components_, Number, is_face>::read_cell_data(
...
+  // access to generic const vectors that have operator ().
+  // FIXME: this is wrong for Trilinos/Petsc MPI vectors
+  // where we should first do Partitioner::local_to_global()
+  template <typename VectorType,
+            typename std::enable_if<!has_local_element<VectorType>::value,
+                                    VectorType>::type * = nullptr>
+  inline typename VectorType::value_type
+  vector_access(const VectorType &vec, const unsigned int entry)
+  {
+    return vec(entry);
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: The question is how we can introduce the local-to-global interface. We probably should pass the partitioner from DoFInfo to this call. Can you open an issue for that so we don't forget it?

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: Done https://github.com/dealii/dealii/issues/7753

**Pull request merged**